Add back the original Keccak submission to SHA-3,

pirapira opened the pull request:
which uses a different padding character.
The tests are taken from revision 3d23e47db69200c2a736c109bb2093213d2d2323

dmentre commented:
Hello @pirapira ,

Several questions on this patch :
- Why do you want to include original Keccak submission?
- Do you have an original reference (e.g. published paper) to original Keccak submission to check padding value?

Review of the patch:
- Its structure is pretty simple
- For new constant "0x06", you need to define it through a symbolic name and define it as C unsigned char
- Same remark for constant "1" used in caml_keccak_extract()
- caml_keccak_extract() seems to be a copy/paste of caml_sha3_extract() so use of OCaml's macros should be OK
- caml_keccak_extract() and caml_sha3_extract() being so similar (except for the padding used), one could ask oneself if better factorization could not be done
- Why "test 98 (hash_extremely_long (Hash.keccak 256))" is commented out?
- How do you have generated the oracles for the test vectors?
- You need to add references to original algorithm and implementation

That's all for my review,
Best regards,
david

pirapira commented:
Thank you very much for a detailed, accurate review.  I'm on vacation until 7 May.  Let me update this PR after that. I'll try to take in all of your comments.

> Why do you want to include original Keccak submission?

When I updated `cryptokit` to the current version, the behavior of the `sha3` hash functions [changed](https://github.com/xavierleroy/cryptokit/blob/2f1858ab3237de3788cae48bc8749777f2ec04bc/Changes#L11), and this affected my project.

I'm trying to build a compiler targeting the Ethereum Virtual Machine. There, the common ABI involves taking the Keccak hash (the version added in this PR) of function signature strings.

> Do you have an original reference (e.g. published paper)

Original, yes (but not in a journal or proceedings).  On page 9 of [Keccak Implementation Overview (Version 3.2)](http://keccak.noekeon.org/Keccak-implementation-3.2.pdf), there is a figure showing `0x01` as the relevant padding.  In a [similar, updated description](http://keccak.noekeon.org/specs_summary.html), at the end of the page, `0x06` is shown as the relevant padding.

There is a more concise description of the change, though not official or original. https://github.com/mjosaarinen/tiny_sha3#updated-07-aug-15

> Why "test 98 (hash_extremely_long (Hash.keccak 256))" is commented out?

The only reason is that I found it commented out in revision 3d23e47 of this repository.  I tried putting it back, and there were no problems.

> How do you have generated the oracles for the test vectors?

The test vectors come from an older revision of this repository, as [told above](https://github.com/xavierleroy/cryptokit/pull/2#issue-225031341).

dmentre commented:
Hello Hirai-san,

Le 2017-04-30 à 20:15, Yoichi Hirai a écrit :
> Thank you very much for a detailed, accurate review. I'm on vacation
> until 7 May. Let me update this PR after that. I'll try to take in all
> of your comments.

OK, let us know when you have more information. There is no hurry.


>     Why do you want to include original Keccak submission?
> 
> When I updated |cryptokit| to the current version, the behavior of the
> |sha3| hash functions changed
> <https://github.com/xavierleroy/cryptokit/blob/2f1858ab3237de3788cae48bc8749777f2ec04bc/Changes#L11>,
> and this affected my project.
> 
> I'm trying to build a compiler targeting the Ethereum Virtual Machine.
> There, the common ABI involves taking the Keccak hash (the version added
> in this PR) of function signature strings.

That's OK for me, but this is up to @xavierleroy to decide to include or
not your patch.

Best regards,
david

pirapira commented:
@dmentre I updated this PR for your comments.

xavierleroy commented:
Thanks for the suggestion and the code.  I agree that having the original Keccak implementation available is useful.  Merging this pull request right now.
